core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/start/StackStartStopService.java

```diff
@@ -70,9 +84,18 @@ public void finishStackStop(StackStartStopContext context) {
         }
     }
 
-    public void handleStackStopError(StackStartStopContext context, StopInstancesResult payload) {
-        LOGGER.error("Error during Stack stop flow:", payload.getErrorDetails());
-        flowMessageService.fireEventAndLog(context.getStack().getId(), Msg.STACK_INFRASTRUCTURE_STOP_FAILED, Status.AVAILABLE.name());
+    public void handleStackStopError(Stack stack, FlowFailureEvent payload) {
+        Exception ex = payload.getException();
+        LOGGER.error("Error during Stack stop flow:", ex);
+        stackUpdater.updateStackStatus(payload.getStackId(), STOP_FAILED, "Stop failed: " + ex.getMessage());
+        flowMessageService.fireEventAndLog(payload.getStackId(), Msg.STACK_INFRASTRUCTURE_STOP_FAILED, STOP_FAILED.name(), ex.getMessage());
+        if (stack.getCluster() != null) {
+            clusterService.updateClusterStatusByStackId(payload.getStackId(), STOPPED);
+            if (stack.getCluster().getEmailNeeded()) {
+                emailSenderService.sendStopFailureEmail(stack.getCluster().getOwner(), stack.getAmbariIp(), stack.getCluster().getName());
+                flowMessageService.fireEventAndLog(payload.getStackId(), Msg.STACK_NOTIFICATION_EMAIL, STOP_FAILED.name());
+            }
+        }
     }
```

Comment:
handleStackStartError and handleStackStopError are containing code duplication, maybe it is eliminate-able.